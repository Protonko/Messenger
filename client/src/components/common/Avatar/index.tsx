import {useMemo, FC} from 'react'
import classNames from 'classnames'
import {Sizes} from 'models/common/sizes'
import {ColorGeneratorByName} from 'utils/ColorGeneratorByName'
import {isDarkColor} from 'utils/isDarkColor'
import {Loader} from 'components/common/Loader'

export interface IPropsAvatar {
  src: string
  name?: string
  size?: Sizes
  customStyles?: string
}

export const Avatar: FC<IPropsAvatar> = ({
  src,
  name,
  size = Sizes.MEDIUM,
  customStyles,
}) => {
  const backgroundColor = useMemo(
    () => new ColorGeneratorByName(name ?? '').generate(),
    [src, name],
  )
  const avatarClassname = classNames(
    'avatar',
    {'avatar--sm': size === Sizes.SMALL},
    {'avatar--lg': size === Sizes.LARGE},
    {[customStyles ?? '']: customStyles},
  )

  if (!src && name) {
    const classNamesText = classNames(
      'avatar__text',
      {'avatar__text--light': isDarkColor(backgroundColor)},
      {'avatar__text--dark': !isDarkColor(backgroundColor)},
    )

    return (
      <div className={avatarClassname} style={{backgroundColor}}>
        <span className={classNamesText}>{name.charAt(0)}</span>
      </div>
    )
  }

  return (
    <div className={avatarClassname}>
      <img className="avatar__image" src={src} alt="" />
      <Loader />
    </div>
  )
}
