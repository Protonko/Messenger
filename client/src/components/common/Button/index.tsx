import {FC, HTMLProps, ReactNode} from 'react'

interface IButtonProps extends HTMLProps<HTMLButtonElement> {
  text?: string
  icon?: ReactNode
  leftAddons?: ReactNode
  rightAddons?: ReactNode
  type?: 'button' | 'submit' | 'reset'
}

export const Button: FC<IButtonProps> = ({
  text,
  icon,
  leftAddons,
  rightAddons,
  ...buttonProps
}) => {
  const renderButtonContent = () => {
    return (
      <>
        {text && (
          <span className="button__text">
            {text}
          </span>
        )}
        {icon && (
          <span className="button__icon">
            {icon}
          </span>
        )}
      </>
    )
  }

  return (
    <button className="button" {...buttonProps}>
      {leftAddons}
      {renderButtonContent()}
      {rightAddons}
    </button>
  )
}
