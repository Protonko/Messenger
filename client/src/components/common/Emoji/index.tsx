import 'emoji-mart/css/emoji-mart.css'
import {FC, useState, useRef} from 'react'
import {EmojiData, Picker} from 'emoji-mart'
import {ReactComponent as Smile} from 'assets/icons/smile.svg'
import {useOutsideClick} from 'hooks/useOutsideClick'

export interface IPropsEmoji {
  onSelect: (emoji: EmojiData) => void
}

export const Emoji: FC<IPropsEmoji> = ({onSelect}) => {
  const [open, setOpen] = useState(false)
  const ref = useRef(null)
  const initiatorRef = useRef(null)

  useOutsideClick(ref, () => setOpen(false), initiatorRef)

  const onSelectEmoji = (emoji: EmojiData) => {
    onSelect(emoji)
    setOpen(false)
  }

  return (
    <div className="emoji">
      {open && (
        <div className="emoji__popup" ref={ref}>
          <Picker native={true} onSelect={onSelectEmoji} />
        </div>
      )}

      <Smile
        className="emoji__button"
        ref={initiatorRef}
        onClick={() => setOpen(!open)}
      />
    </div>
  )
}
