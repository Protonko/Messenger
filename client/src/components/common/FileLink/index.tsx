import type {FC} from 'react'
import {ReactComponent as FileIcon} from 'assets/icons/file.svg'

export interface IFileLinkProps {
  link: string
}

export const FileLink: FC<IFileLinkProps> = ({link}) => {
  return (
    <a download href={link} className="file-link">
      <FileIcon className="file-link__icon" />
      <div className="file-link__info">
        <span className="file-link__info-name">{link.split('/').pop()}</span>
      </div>
    </a>
  )
}
