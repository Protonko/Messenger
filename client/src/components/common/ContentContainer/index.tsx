import type {FC} from 'react'
import {Loader} from 'components/common/Loader'
import {Text} from 'components/common/Text'

interface ContentContainerProps {
  loading: boolean
  errorMessage: string | null
  children: JSX.Element
}

export const ContentContainer: FC<ContentContainerProps> = ({
  loading,
  errorMessage,
  children,
}) => {
  if (errorMessage) {
    return <Text>{errorMessage}</Text>
  }

  if (loading) {
    return <Loader center={true} />
  }

  return children
}
