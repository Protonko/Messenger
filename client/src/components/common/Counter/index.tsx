// types
import {Status} from 'models/common/status'

import {FC} from 'react'
import classNames from 'classnames'

export interface IPropsCounter {
  count: number | '99+',
  status?: Status,
}

export const Counter: FC<IPropsCounter> = ({
  count,
  status = Status.ACTIVE,
}) => {
  const classNameStatus = classNames(
    'counter',
    {'counter--muted': status === Status.MUTED}
  )

  return (
    <div className={classNameStatus}>
      <span className="counter__text">
        {count}
      </span>
    </div>
  )
}
