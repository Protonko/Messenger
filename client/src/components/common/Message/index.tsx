import type {FC} from 'react'
import type {IMessage} from 'models/message'
import classNames from 'classnames'
import {TextSize, TextTypes} from 'models/common/text'
import {Sizes} from 'models/common/sizes'
import {Avatar} from 'components/common/Avatar'
import {Text} from 'components/common/Text'

export interface IPropsMessage extends IMessage {
  customStyles?: string
  avatarSize?: Sizes
}

export const Message: FC<IPropsMessage> = ({
  customStyles,
  text,
  author,
  createdAt,
  updatedAt,
  attachments,
}) => {
  const classNamesMessage = classNames('message', {
    [customStyles ?? '']: !!customStyles,
  })

  const renderText = () => {
    if (text) {
      return (
        <Text
          type={TextTypes.mixed}
          size={TextSize.EXTRA_SMALL}
          customStyles="message__content-item message__content-item--text"
          numberOfLines={1}
        >
          {text}
        </Text>
      )
    }
  }

  const renderAttachments = () => {
    if (attachments.length) {
      return (
        <img
          className="message__content-item message__content-item--image"
          src=""
          alt="#"
        />
      )
    }
  }

  return (
    <div className={classNamesMessage}>
      <div className="message__body">
        <Avatar name={author.full_name} src={author.avatar ?? ''} />

        <div className="message__data">
          <Text
            type={TextTypes.h4}
            customStyles="message__author"
            numberOfLines={1}
          >
            {author.full_name}
          </Text>

          <div className="message__content">
            {renderText()}
            {renderAttachments()}
          </div>
        </div>

        <div className="message__date">
          <Text
            type={TextTypes.mixed}
            size={TextSize.EXTRA_SMALL}
            customStyles="message__date-text"
            numberOfLines={1}
          >
            {new Date(updatedAt || createdAt).toLocaleString()}
          </Text>
        </div>
      </div>
    </div>
  )
}
