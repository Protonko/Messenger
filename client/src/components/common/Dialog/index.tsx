import type {FC} from 'react'
import type {IDialog} from 'models/dialog'
import {ReadStatus} from 'models/common/status'
import {TextSize, TextTypes} from 'models/common/text'
import {ReactComponent as DoubleCheck} from 'assets/icons/double-check.svg'
import {ReactComponent as Check} from 'assets/icons/check.svg'
import {COLORS} from 'static/colors'
import {Text} from 'components/common/Text'
import {Counter} from 'components/common/Counter'
import {Avatar} from 'components/common/Avatar'

export const Dialog: FC<IDialog> = ({
  name,
  description,
  avatar,
  edited,
  createdAt,
  updatedAt,
  messages,
  status,
  readStatus,
}) => {
  const counter = messages.toString().length > 2 ? '99+' : messages

  const renderStatus = () => {
    if (!!counter) {
      return <Counter count={counter} status={status}/>
    }

    if (readStatus) {
      const props = {
        width: 15,
        height: 15,
        color: COLORS.dustyGray,
      }
      return readStatus === ReadStatus.READ ? <DoubleCheck {...props} /> : <Check {...props} />
    }

    return null
  }

  return (
    <div className="dialog">
      <Avatar src={avatar} />

      <div className="dialog__data">
        <div className="dialog__text">
          <Text
            type={TextTypes.h4}
            customStyles="dialog__text-title"
            numberOfLines={1}
          >
            {name}
          </Text>
          <Text>{description}</Text>
        </div>

        <div className="dialog__info">
          <Text
            type={TextTypes.mixed}
            size={TextSize.EXTRA_SMALL}
            customStyles="dialog__info-date"
          >
            {new Date(edited ? updatedAt : createdAt).toLocaleDateString()}
          </Text>

          {renderStatus()}
        </div>
      </div>
    </div>
  )
}
