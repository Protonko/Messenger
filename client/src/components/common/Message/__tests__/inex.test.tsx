import {shallow, ShallowWrapper} from 'enzyme'
import {Message, IPropsMessage} from 'components/common/Message'
import {FileLink} from 'components/common/FileLink'
import {MESSAGE} from '../../../../static/test-mocks/message'

const ELEMENT_SELECTORS = {
  message: '.message',
  text: '.message__content-item--text',
  image: '.message__content-item--image',
  additional: 'additionalClassName',
  selected: '.message--selected',
}

describe('Message', () => {
  let component: ShallowWrapper<IPropsMessage>
  let props: IPropsMessage

  beforeEach(() => {
    props = {
      ...MESSAGE,
    }
  })

  it('Should render Message component', () => {
    component = shallow(<Message {...props} />)

    expect(component.find(ELEMENT_SELECTORS.message).length).toBe(1)
    expect(component).toMatchSnapshot('Message')
  })

  it('Should render Message component with selected modifier', () => {
    props.selected = true
    component = shallow(<Message {...props} />)

    expect(component.find(ELEMENT_SELECTORS.selected).length).toBe(1)
    expect(component).toMatchSnapshot('Selected message')
  })

  it('Should render Message component without text', () => {
    component = shallow(<Message {...props} text="" />)

    expect(component.find(ELEMENT_SELECTORS.text).length).toBe(0)
    expect(component).toMatchSnapshot('Message without text')
  })

  it('Should render Message component with image', () => {
    component = shallow(<Message {...props} attachment="image.png" />)

    expect(component.find(ELEMENT_SELECTORS.image).length).toBe(1)
    expect(component).toMatchSnapshot('Message with image')
  })

  it('Should render Message component with file', () => {
    component = shallow(<Message {...props} attachment="doc.pdf" />)

    expect(component.find(FileLink).length).toBe(1)
    expect(component).toMatchSnapshot('Message with file')
  })

  it('Should render Message component with additional classname', () => {
    component = shallow(
      <Message {...props} additionalClassname={ELEMENT_SELECTORS.additional} />,
    )

    expect(component.find(`.${ELEMENT_SELECTORS.additional}`).length).toBe(1)
    expect(component).toMatchSnapshot('Message with additionalClassName')
  })

  it('Should render Message component with created date', () => {
    component = shallow(<Message {...props} updatedAt="" />)

    expect(component).toMatchSnapshot('Message with created date')
  })
})
