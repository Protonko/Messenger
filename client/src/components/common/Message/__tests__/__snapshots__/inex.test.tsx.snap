// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message Should render Message component with additional classname: Message with additionalClassName 1`] = `
<div
  className="message additionalClassName"
>
  <div
    className="message__body"
  >
    <Avatar
      name="name"
      src=""
    />
    <div
      className="message__data"
    >
      <Component
        customStyles="message__author"
        numberOfLines={1}
        type="h4"
      >
        name
      </Component>
      <div
        className="message__content"
      >
        <Component
          customStyles="message__content-item message__content-item--text"
          numberOfLines={1}
          size="XS"
          type="mixed"
        >
          text
        </Component>
      </div>
    </div>
    <div
      className="message__date"
    >
      <Component
        customStyles="message__date-text"
        numberOfLines={1}
        size="XS"
        type="mixed"
      >
        Invalid Date
      </Component>
    </div>
  </div>
</div>
`;

exports[`Message Should render Message component with created date: Message with created date 1`] = `
<div
  className="message"
>
  <div
    className="message__body"
  >
    <Avatar
      name="name"
      src=""
    />
    <div
      className="message__data"
    >
      <Component
        customStyles="message__author"
        numberOfLines={1}
        type="h4"
      >
        name
      </Component>
      <div
        className="message__content"
      >
        <Component
          customStyles="message__content-item message__content-item--text"
          numberOfLines={1}
          size="XS"
          type="mixed"
        >
          text
        </Component>
      </div>
    </div>
    <div
      className="message__date"
    >
      <Component
        customStyles="message__date-text"
        numberOfLines={1}
        size="XS"
        type="mixed"
      >
        Invalid Date
      </Component>
    </div>
  </div>
</div>
`;

exports[`Message Should render Message component with file: Message with file 1`] = `
<div
  className="message"
>
  <div
    className="message__body"
  >
    <Avatar
      name="name"
      src=""
    />
    <div
      className="message__data"
    >
      <Component
        customStyles="message__author"
        numberOfLines={1}
        type="h4"
      >
        name
      </Component>
      <div
        className="message__content"
      >
        <Component
          customStyles="message__content-item message__content-item--text"
          numberOfLines={1}
          size="XS"
          type="mixed"
        >
          text
        </Component>
        <div
          className="message__content-item"
        >
          <FileLink
            link="doc.pdf"
          />
        </div>
      </div>
    </div>
    <div
      className="message__date"
    >
      <Component
        customStyles="message__date-text"
        numberOfLines={1}
        size="XS"
        type="mixed"
      >
        Invalid Date
      </Component>
    </div>
  </div>
</div>
`;

exports[`Message Should render Message component with image: Message with image 1`] = `
<div
  className="message"
>
  <div
    className="message__body"
  >
    <Avatar
      name="name"
      src=""
    />
    <div
      className="message__data"
    >
      <Component
        customStyles="message__author"
        numberOfLines={1}
        type="h4"
      >
        name
      </Component>
      <div
        className="message__content"
      >
        <Component
          customStyles="message__content-item message__content-item--text"
          numberOfLines={1}
          size="XS"
          type="mixed"
        >
          text
        </Component>
        <img
          alt="image.png"
          className="message__content-item message__content-item--image"
          src="image.png"
        />
      </div>
    </div>
    <div
      className="message__date"
    >
      <Component
        customStyles="message__date-text"
        numberOfLines={1}
        size="XS"
        type="mixed"
      >
        Invalid Date
      </Component>
    </div>
  </div>
</div>
`;

exports[`Message Should render Message component without text: Message without text 1`] = `
<div
  className="message"
>
  <div
    className="message__body"
  >
    <Avatar
      name="name"
      src=""
    />
    <div
      className="message__data"
    >
      <Component
        customStyles="message__author"
        numberOfLines={1}
        type="h4"
      >
        name
      </Component>
      <div
        className="message__content"
      />
    </div>
    <div
      className="message__date"
    >
      <Component
        customStyles="message__date-text"
        numberOfLines={1}
        size="XS"
        type="mixed"
      >
        Invalid Date
      </Component>
    </div>
  </div>
</div>
`;

exports[`Message Should render Message component: Message 1`] = `
<div
  className="message"
>
  <div
    className="message__body"
  >
    <Avatar
      name="name"
      src=""
    />
    <div
      className="message__data"
    >
      <Component
        customStyles="message__author"
        numberOfLines={1}
        type="h4"
      >
        name
      </Component>
      <div
        className="message__content"
      >
        <Component
          customStyles="message__content-item message__content-item--text"
          numberOfLines={1}
          size="XS"
          type="mixed"
        >
          text
        </Component>
      </div>
    </div>
    <div
      className="message__date"
    >
      <Component
        customStyles="message__date-text"
        numberOfLines={1}
        size="XS"
        type="mixed"
      >
        Invalid Date
      </Component>
    </div>
  </div>
</div>
`;
